src/state/routing/utils.ts

```diff
@@ -89,6 +114,10 @@ const parsePair = ({ reserve0, reserve1 }: V2PoolInRoute): Pair =>
     CurrencyAmount.fromRawAmount(parseToken(reserve1.token), reserve1.quotient)
   )
 
-function isV3Route(route: V3PoolInRoute[] | V2PoolInRoute[]): route is V3PoolInRoute[] {
-  return route[0].type === 'v3-pool'
+function isV2OnlyRoute(route: (V3PoolInRoute | V2PoolInRoute)[]): route is V2PoolInRoute[] {
```

Comment:
nit: Can you replace both of these with a generic fn, because they essentially disambiguate a generic type?

    const isOnlyV2 = isVersionedRoute<V2PoolInRoute>(PoolType.V2Pool, route)
    const isOnlyV3 = isVersionedRoute<V3PoolInRoute>(PoolType.V3Pool, route)

    function isVersionedRoute<T extends V2PoolInRoute | V3PoolInRoute>(
      type: T['type'],
      route: (V3PoolInRoute | V2PoolInRoute)[]
    ): route is T[] {
      return route.every((pool) => pool.type === type)
    }

Reply:
good idea!